# Lab | Data Aggregation and Filtering

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by first performing data cleaning, formatting, and structuring.

In [1]:
#Starting by importing the libraries, functions and loading the CSV file:

import pandas as pd
import functions

df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv")

display(df)

Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
0               0  DK49336     Arizona              4809.216960       No   
1               1  KX64629  California              2228.525238       No   
2               2  LZ68649  Washington             14947.917300       No   
3               3  XL78013      Oregon             22332.439460      Yes   
4               4  QA50777      Oregon              9025.067525       No   
...           ...      ...         ...                      ...      ...   
10905       10905  FE99816      Nevada             15563.369440       No   
10906       10906  KX53892      Oregon              5259.444853       No   
10907       10907  TL39050     Arizona             23893.304100       No   
10908       10908  WA60547  California             11971.977650       No   
10909       10909  IV32877         NaN              6857.519928      NaN   

       Coverage Education Effective To Date EmploymentStatus Gender  ...  \
0         Basic   College           2/18/11         Employed      M  ...   
1         Basic   College           1/18/11       Unemployed      F  ...   
2         Basic  Bachelor           2/10/11         Employed      M  ...   
3      Extended   College           1/11/11         Employed      M  ...   
4       Premium  Bachelor           1/17/11    Medical Leave      F  ...   
...         ...       ...               ...              ...    ...  ...   
10905   Premium  Bachelor           1/19/11       Unemployed      F  ...   
10906     Basic   College            1/6/11         Employed      F  ...   
10907  Extended  Bachelor            2/6/11         Employed      F  ...   
10908   Premium   College           2/13/11         Employed      F  ...   
10909     Basic  Bachelor            1/8/11       Unemployed      M  ...   

       Number of Open Complaints Number of Policies     Policy Type  \
0                            0.0                  9  Corporate Auto   
1                            0.0                  1   Personal Auto   
2                            0.0                  2   Personal Auto   
3                            0.0                  2  Corporate Auto   
4                            NaN                  7   Personal Auto   
...                          ...                ...             ...   
10905                        NaN                  7   Personal Auto   
10906                        0.0                  6   Personal Auto   
10907                        0.0                  2  Corporate Auto   
10908                        4.0                  6   Personal Auto   
10909                        0.0                  3   Personal Auto   

             Policy  Renew Offer Type  Sales Channel  Total Claim Amount  \
0      Corporate L3            Offer3          Agent          292.800000   
1       Personal L3            Offer4    Call Center          744.924331   
2       Personal L3            Offer3    Call Center          480.000000   
3      Corporate L3            Offer2         Branch          484.013411   
4       Personal L2            Offer1         Branch          707.925645   
...             ...               ...            ...                 ...   
10905   Personal L1            Offer3            Web         1214.400000   
10906   Personal L3            Offer2         Branch          273.018929   
10907  Corporate L3            Offer1            Web          381.306996   
10908   Personal L1            Offer1         Branch          618.288849   
10909   Personal L1            Offer4            Web         1021.719397   

       Vehicle Class Vehicle Size Vehicle Type  
0      Four-Door Car      Medsize          NaN  
1      Four-Door Car      Medsize          NaN  
2                SUV      Medsize            A  
3      Four-Door Car      Medsize            A  
4      Four-Door Car      Medsize          NaN  
...              ...          ...          ...  
10905     Luxury Car      Medsize            A  
10906  Four-Door Car   

In [2]:
# Cleaning the column names and dropping the 'unnamed:_0' column which was a copy of the label column
df=functions.cleaning_column_names(df)
df=df.drop('unnamed:_0', axis=1)

df.head()

customer       state  customer_lifetime_value response  coverage education  \
0  DK49336     Arizona              4809.216960       No     Basic   College   
1  KX64629  California              2228.525238       No     Basic   College   
2  LZ68649  Washington             14947.917300       No     Basic  Bachelor   
3  XL78013      Oregon             22332.439460      Yes  Extended   College   
4  QA50777      Oregon              9025.067525       No   Premium  Bachelor   

  effective_to_date employmentstatus gender  income  ...  \
0           2/18/11         Employed      M   48029  ...   
1           1/18/11       Unemployed      F       0  ...   
2           2/10/11         Employed      M   22139  ...   
3           1/11/11         Employed      M   49078  ...   
4           1/17/11    Medical Leave      F   23675  ...   

  number_of_open_complaints number_of_policies     policy_type        policy  \
0                       0.0                  9  Corporate Auto  Corporate L3   
1                       0.0                  1   Personal Auto   Personal L3   
2                       0.0                  2   Personal Auto   Personal L3   
3                       0.0                  2  Corporate Auto  Corporate L3   
4                       NaN                  7   Personal Auto   Personal L2   

   renew_offer_type  sales_channel  total_claim_amount  vehicle_class  \
0            Offer3          Agent          292.800000  Four-Door Car   
1            Offer4    Call Center          744.924331  Four-Door Car   
2            Offer3    Call Center          480.000000            SUV   
3            Offer2         Branch          484.013411  Four-Door Car   
4            Offer1         Branch          707.925645  Four-Door Car   

  vehicle_size vehicle_type  
0      Medsize          NaN  
1      Medsize          NaN  
2      Medsize            A  
3      Medsize            A  
4      Medsize          NaN  

[5 rows x 25 columns]

In [3]:
#Checking for duplicates and cleaning them:

display(df.duplicated().sum())
df=functions.dealing_with_duplicates(df)
display(df.duplicated().sum())

37

0

In [4]:
#Checking for entire rows of missing values:
df.isna().all(axis=1).sum()

0

In [5]:
#Checking for percentage of missing values in each column:

display(df.isna().sum()/len(df)*100)

customer                          0.000000
state                             5.647016
customer_lifetime_value           0.000000
response                          5.647016
coverage                          0.000000
education                         0.000000
effective_to_date                 0.000000
employmentstatus                  0.000000
gender                            0.000000
income                            0.000000
location_code                     0.000000
marital_status                    0.000000
monthly_premium_auto              0.000000
months_since_last_claim           5.729789
months_since_policy_inception     0.000000
number_of_open_complaints         5.729789
number_of_policies                0.000000
policy_type                       0.000000
policy                            0.000000
renew_offer_type                  0.000000
sales_channel                     0.000000
total_claim_amount                0.000000
vehicle_class                     5.591833
vehicle_siz

In [6]:
#Checking the column data types.
display(df.dtypes)

df['effective_to_date']=pd.to_datetime(df['effective_to_date'])

display(df.dtypes)

customer                          object
state                             object
customer_lifetime_value          float64
response                          object
coverage                          object
education                         object
effective_to_date                 object
employmentstatus                  object
gender                            object
income                             int64
location_code                     object
marital_status                    object
monthly_premium_auto               int64
months_since_last_claim          float64
months_since_policy_inception      int64
number_of_open_complaints        float64
number_of_policies                 int64
policy_type                       object
policy                            object
renew_offer_type                  object
sales_channel                     object
total_claim_amount               float64
vehicle_class                     object
vehicle_size                      object
vehicle_type    

customer                                 object
state                                    object
customer_lifetime_value                 float64
response                                 object
coverage                                 object
education                                object
effective_to_date                datetime64[ns]
employmentstatus                         object
gender                                   object
income                                    int64
location_code                            object
marital_status                           object
monthly_premium_auto                      int64
months_since_last_claim                 float64
months_since_policy_inception             int64
number_of_open_complaints               float64
number_of_policies                        int64
policy_type                              object
policy                                   object
renew_offer_type                         object
sales_channel                           

In [7]:
# Creating new functions to clean the missing values:

def fill_nan_median(dataframe: pd.DataFrame, col_name: str)-> pd.DataFrame:
    dataframe2=dataframe.copy()
    dataframe2[col_name]=dataframe2[col_name].fillna(dataframe2[col_name].median())
    return dataframe2

def fill_nan_mean(dataframe: pd.DataFrame, col_name: str)-> pd.DataFrame:
    dataframe2=dataframe.copy()
    dataframe2[col_name]=dataframe2[col_name].fillna(dataframe2[col_name].mean())
    return dataframe2

def fill_nan_mode(dataframe: pd.DataFrame, col_name: str)-> pd.DataFrame:
    dataframe2=dataframe.copy()
    dataframe2[col_name]=dataframe2[col_name].fillna(dataframe2[col_name].mode()[0])
    return dataframe2

In [8]:
#In the categorical columns I'm' going to fill the missing values with the mode:

df = fill_nan_mode(df, "state")
df = fill_nan_mode(df, "response")
df = fill_nan_mode(df, "vehicle_class")
df = fill_nan_mode(df, "vehicle_size")
df = fill_nan_mode(df, "vehicle_type")

display(df.isna().sum()/len(df)*100)

customer                         0.000000
state                            0.000000
customer_lifetime_value          0.000000
response                         0.000000
coverage                         0.000000
education                        0.000000
effective_to_date                0.000000
employmentstatus                 0.000000
gender                           0.000000
income                           0.000000
location_code                    0.000000
marital_status                   0.000000
monthly_premium_auto             0.000000
months_since_last_claim          5.729789
months_since_policy_inception    0.000000
number_of_open_complaints        5.729789
number_of_policies               0.000000
policy_type                      0.000000
policy                           0.000000
renew_offer_type                 0.000000
sales_channel                    0.000000
total_claim_amount               0.000000
vehicle_class                    0.000000
vehicle_size                     0

In [9]:
#Checking median and mean of the numeric columns:

display(df['months_since_last_claim'].median())
display(df['months_since_last_claim'].mean())

display(df['number_of_open_complaints'].median())
display(df['number_of_open_complaints'].mean())

14.0

15.145073170731708

0.0

0.3835121951219512

In [10]:
#Since these are supposed to be whole numbers i prefer to use the median to fill the missing values:

df = fill_nan_median(df, "months_since_last_claim")

df = fill_nan_median(df, "number_of_open_complaints")

display(df.isna().sum()/len(df)*100)

customer                         0.0
state                            0.0
customer_lifetime_value          0.0
response                         0.0
coverage                         0.0
education                        0.0
effective_to_date                0.0
employmentstatus                 0.0
gender                           0.0
income                           0.0
location_code                    0.0
marital_status                   0.0
monthly_premium_auto             0.0
months_since_last_claim          0.0
months_since_policy_inception    0.0
number_of_open_complaints        0.0
number_of_policies               0.0
policy_type                      0.0
policy                           0.0
renew_offer_type                 0.0
sales_channel                    0.0
total_claim_amount               0.0
vehicle_class                    0.0
vehicle_size                     0.0
vehicle_type                     0.0
dtype: float64

1. Create a new DataFrame that only includes customers who have a total_claim_amount greater than $1,000 and have a response of "Yes" to the last marketing campaign.

In [11]:
df_filtered=df[(df["total_claim_amount"] > 1000) & (df["response"] == "Yes")].reset_index()
df_filtered=df_filtered.drop('index', axis=1)
df_filtered

customer       state  customer_lifetime_value response  coverage  \
0   OK31456  California             11009.130490      Yes   Premium   
1   YJ16163      Oregon             11009.130490      Yes   Premium   
2   GW43195      Oregon             25807.063000      Yes  Extended   
3   IP94270     Arizona             13736.132500      Yes   Premium   
4   FJ28407  California              5619.689084      Yes   Premium   
..      ...         ...                      ...      ...       ...   
62  FN44127      Oregon              3508.569533      Yes  Extended   
63  XZ64172      Oregon             10963.957230      Yes   Premium   
64  IX60941      Oregon              3508.569533      Yes  Extended   
65  QO62792      Oregon              7840.165778      Yes  Extended   
66  CK39096      Oregon              5619.689084      Yes   Premium   

               education effective_to_date employmentstatus gender  income  \
0               Bachelor        2011-01-24         Employed      F   51643   
1               Bachelor        2011-01-24         Employed      F   51643   
2                College        2011-02-13         Employed      F   71210   
3                 Master        2011-02-13         Disabled      F   16181   
4   High School or Below        2011-01-26       Unemployed      M       0   
..                   ...               ...              ...    ...     ...   
62               College        2011-01-05    Medical Leave      M   20978   
63  High School or Below        2011-02-08         Employed      M   55687   
64               College        2011-01-05    Medical Leave      M   20978   
65               College        2011-01-14         Employed      M   58414   
66  High School or Below        2011-01-26       Unemployed      M       0   

    ... number_of_open_complaints number_of_policies     policy_type  \
0   ...                       0.0                  1  Corporate Auto   
1   ...                       0.0                  1    Special Auto   
2   ...                       1.0                  2   Personal Auto   
3   ...                       0.0                  8   Personal Auto   
4   ...                       0.0                  1   Personal Auto   
..  ...                       ...                ...             ...   
62  ...                       1.0                  1   Personal Auto   
63  ...                       0.0                  1  Corporate Auto   
64  ...                       1.0                  1   Personal Auto   
65  ...                       2.0                  1   Personal Auto   
66  ...                       0.0                  1   Personal Auto   

          policy  renew_offer_type  sales_channel  total_claim_amount  \
0   Corporate L3            Offer2          Agent         1358.400000   
1     Special L3            Offer2          Agent         1358.400000   
2    Personal L2            Offer1         Branch         1027.200000   
3    Personal L2            Offer1            Web         1261.319869   
4    Personal L1            Offer2            Web         1027.000029   
..           ...               ...            ...                 ...   
62   Personal L2            Offer2         Branch         1176.278800   
63  Corporate L2            Offer1          Agent         1324.800000   
64   Personal L3            Offer2         Branch         1176.278800   
65   Personal L3            Offer2          Agent         1008.000000   
66   Personal L3            Offer2            Web         1027.000029   

    vehicle_class vehicle_size vehicle_type  
0      Luxury Car      Medsize            A  
1      Luxury Car      Medsize            A  
2      Luxury Car        Small            A  
3             SUV      Medsize            A  
4             SUV      Medsize            A  
..            ...          ...          ...  
62  Four-Door Car        Small            A  
63     Luxury SUV      Medsize            A  
64  Four-Door Car        Small            A  
65  Four-Do

2. Using the original Dataframe, analyze the average total_claim_amount by each policy type and gender for customers who have responded "Yes" to the last marketing campaign. Write your conclusions. 

In [12]:
df2= df[df["response"] == "Yes"].groupby(['policy_type','gender','response']).agg({"total_claim_amount": 'mean'}).reset_index()

display(df2)

policy_type gender response  total_claim_amount
0  Corporate Auto      F      Yes          433.738499
1  Corporate Auto      M      Yes          407.829300
2   Personal Auto      F      Yes          452.755476
3   Personal Auto      M      Yes          457.010178
4    Special Auto      F      Yes          453.280164
5    Special Auto      M      Yes          429.527942

##### Answer:

The highest average total claim amount for customers who have responded "Yes" to the last marketing campaign is from males with Personal Auto policy type with 457.010178.
The lowest average was for males with Corporate Auto policy type with 407.829300.

3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

In [13]:
df_customers_by_state = df.groupby(['state']).agg({"customer": "count"}).reset_index()

display(df_customers_by_state)

df_customers_by_state_filtered = df_customers_by_state[df_customers_by_state['customer'] > 500]

df_customers_by_state_filtered

state  customer
0     Arizona      1934
1  California      4162
2      Nevada       992
3      Oregon      2897
4  Washington       888

state  customer
0     Arizona      1934
1  California      4162
2      Nevada       992
3      Oregon      2897
4  Washington       888

4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.

In [14]:
df.groupby(['education','gender']).agg({"customer_lifetime_value": [max, min,'median']}).reset_index()

education gender customer_lifetime_value               \
                                                   max          min   
0              Bachelor      F             73225.95652  1904.000852   
1              Bachelor      M             67907.27050  1898.007675   
2               College      F             61850.18803  1898.683686   
3               College      M             61134.68307  1918.119700   
4                Doctor      F             44856.11397  2395.570000   
5                Doctor      M             32677.34284  2267.604038   
6  High School or Below      F             55277.44589  2144.921535   
7  High School or Below      M             83325.38119  1940.981221   
8                Master      F             51016.06704  2417.777032   
9                Master      M             50568.25912  2272.307310   

                
        median  
0  5632.613078  
1  5548.031892  
2  5623.611187  
3  6005.847375  
4  5332.462694  
5  5577.669457  
6  6035.086336  
7  6286.731006  
8  5729.855012  
9  5579.099207

##### Conclusions:

From the data we can observe that females with a Bachelor degree have the highest customer lifetime value (73225.95652) and the lowest is for males with a Bachelor Degree (1898.007675). The highest median value is for males with a high school or below level of education (6286.731006) and the lowest median is females with a doctor's degree (5332.462694).

## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

In [105]:
df2=df.copy()

df2['month'] = pd.to_datetime(df['effective_to_date'], format='%m').dt.month_name()

grouped_table=df2.groupby(['state','month']).agg({"number_of_policies": ['count']}).reset_index()

grouped_table

pivot_table=pd.pivot(grouped_table, index='state', columns='month').reset_index()
display(pivot_table)

pivot_table.columns = pivot_table.columns.get_level_values(1)
pivot_table.columns.values[0] = 'state'
pivot_table.columns.values[1] = 'February'
pivot_table.columns.values[2] = 'January'
pivot_table

state number_of_policies        
                               count        
month                       February January
0         Arizona                927    1007
1      California               1941    2221
2          Nevada                441     551
3          Oregon               1340    1557
4      Washington                425     463

state  February  January
0     Arizona       927     1007
1  California      1941     2221
2      Nevada       441      551
3      Oregon      1340     1557
4  Washington       425      463

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.* !SORT_VALUES (BY=['col1','col2'], ASCENDING=[FALSE, TRUE])!
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

In [116]:
df3=df.copy()

df3['month'] = pd.to_datetime(df['effective_to_date'], format='%m').dt.month_name()

# To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group.

df3=df3.groupby(['state','month']).agg({"number_of_policies": ['count']})

# Afterwards, you will need to sort the data by the count of policies sold in descending order.

df3=df3.sort_values(by=[('number_of_policies', 'count')], ascending=False).reset_index()

#Next, you will select the top 3 states with the highest number of policies sold.

df3=df3[(df3['state'] == 'California') | (df3['state'] == 'Oregon') | (df3['state'] == 'Arizona')]

df3

state     month number_of_policies
                                     count
0  California   January               2221
1  California  February               1941
2      Oregon   January               1557
3      Oregon  February               1340
4     Arizona   January               1007
5     Arizona  February                927

7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

In [196]:
# your code goes here
df4=df.copy()
#Creating a new df where I filter the responses with "Yes" as condition, and I group the column sales channels and add as values the total "Yes" reponses:
df5=df4[df4["response"] == "Yes"].groupby('sales_channel').agg({"response": 'count'}).reset_index()

#Creating a new as above but this time without filtering the responses, so as to get the total count of reponses dividied by sales channel:
df6=df4.groupby('sales_channel').agg({"response": 'count'}).reset_index()
display(df5)
display(df6)
#Creating a new column "response_rate" with the rate from the "Yes" responses divided by the total by sales channel:
df6['response_rate']=df5['response'] / df6['response']
df6

sales_channel  response
0         Agent       741
1        Branch       325
2   Call Center       221
3           Web       177

sales_channel  response
0         Agent      4111
1        Branch      3013
2   Call Center      2132
3           Web      1617

sales_channel  response  response_rate
0         Agent      4111       0.180248
1        Branch      3013       0.107866
2   Call Center      2132       0.103659
3           Web      1617       0.109462